src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -398,142 +340,78 @@ private void failUnpublished(final List<BatchItem> batch, final String reason) {
+    public EventConsumer createEventConsumer(final List<NakadiCursor> cursors)
             throws ServiceUnavailableException, InvalidCursorException {
-        this.validateCursors(topic, cursors);
-
-        final List<KafkaCursor> kafkaCursors = Lists.newArrayListWithCapacity(cursors.size());
-
-        for (final Cursor cursor : cursors) {
-            final String offset = cursor.getOffset();
-            final String partition = cursor.getPartition();
-
-            final long kafkaOffset;
-            if (Cursor.BEFORE_OLDEST_OFFSET.equals(offset)) {
-                final TopicPartition tp = getPartition(topic, partition);
-                kafkaOffset = toKafkaOffset(tp.getOldestAvailableOffset());
-            } else {
-                kafkaOffset = toKafkaOffset(offset) + 1L;
-            }
-
-            final KafkaCursor kafkaCursor = kafkaCursor(toKafkaPartition(partition), kafkaOffset);
-            kafkaCursors.add(kafkaCursor);
-        }
-
-        return kafkaFactory.createNakadiConsumer(topic, kafkaCursors, nakadiSettings.getKafkaPollTimeoutMs());
+        return kafkaFactory.createNakadiConsumer(
+                this.validateConsumerCursors(cursors),
+                nakadiSettings.getKafkaPollTimeoutMs());
     }
 
-    public int compareOffsets(final String firstOffset, final String secondOffset) {
+    public int compareOffsets(final NakadiCursor first, final NakadiCursor second) {
         try {
-            final long first = toKafkaOffset(firstOffset);
-            final long second = toKafkaOffset(secondOffset);
-            return Long.compare(first, second);
-        } catch (final NumberFormatException e) {
+            return KafkaCursor.fromNakadiPosition(first).compareTo(KafkaCursor.fromNakadiPosition(second));
+        } catch (final InvalidCursorException e) {
             throw new IllegalArgumentException("Incorrect offset format, should be long", e);
         }
     }
 
-    private void validateCursors(final String topic, final List<Cursor> cursors) throws ServiceUnavailableException,
-            InvalidCursorException {
-        final List<TopicPartition> partitions = listPartitions(topic);
-
-        for (final Cursor cursor : cursors) {
-            validateCursorForNulls(cursor);
-
-            final TopicPartition topicPartition = partitions
-                    .stream()
-                    .filter(tp -> tp.getPartitionId().equals(cursor.getPartition()))
-                    .findFirst()
-                    .orElseThrow(() -> new InvalidCursorException(PARTITION_NOT_FOUND, cursor));
-
-            if (Cursor.BEFORE_OLDEST_OFFSET.equals(cursor.getOffset())) {
-                continue;
-            } else if (Cursor.BEFORE_OLDEST_OFFSET.equals(topicPartition.getNewestAvailableOffset())) {
-                throw new InvalidCursorException(EMPTY_PARTITION, cursor);
+    private List<KafkaCursor> validateConsumerCursors(final List<NakadiCursor> cursors)
```

Comment:
@lmontrieux No, there's actually no difference between Nakadi and Kafka cursor. The event having offset 43248 in Kafka will have offset "43248" in Nakadi. The difference is in the reading semantics. In Kafka it is: "give me messages starting from offset X". In Nakadi it is "give me messages starting from event next after X". That's why we need to add 1 when we start reading in Kafka.